### PR TITLE
Correctly parse and compare JSON docs with numbers that are too large to fit in a float64 without losing precision.

### PR DESCRIPTION
**Pull request by nicktobey**
Previously we would parse all numbers in serialized JSON as 64-bit floats, and all comparisons between numbers in JSON would be coerced to 64-bit floats.

This was causing problems for inputs that can't be represented precisely as a float, but can be represented precisely as an int64 or uint64.

This PR enhances the logic for both parsing and comparing JSON documents to correctly handle float64, int64, and uint64 without any loss of precision.

### CODE REVIEW

File: sql/types/json.go [hunk 1]
```diff
@@ -232,8 +238,68 @@ func DeepCopyJson(v interface{}) interface{} {
...
+// for each number to avoid losing precision whenever possible.
+func JsonUnmarshal(data []byte, v *interface{}) error {
+	dec := json.NewDecoder(bytes.NewReader(data))
+	dec.UseNumber()
+	if err := dec.Decode(v); err != nil {
+		return err
+	}
+
+	// Unlike json.Unmarshal, we need to check that the decoder has no more tokens
+	if _, err := dec.Token(); err != io.EOF {
```
Comment (zachmu) [inline sql/types/json.go]: Shouldn't you also check token is nil if we're trying to validate this is valid JSON?

File: enginetest/queries/queries.go [hunk 2]
```diff
@@ -8726,6 +8726,26 @@ from typestable`,
 			{"false"},
 		},
 	},
+	{
+		// 9223372036854775808 cannot be represented as a float64
+		Query: "select cast('-9223372036854775806' as json);",
```
Comment (zachmu) [inline enginetest/queries/queries.go:8731]: Probably want to put some engine tests in a json-specific location to make sure these edge case numbers work correctly for:

* Storage and retrieval of docs
* Values in arrays and maps

**Review by zachmu — APPROVED** [review]
LGTM

In addition to these comments, it's probably worthwhile to write a couple backwards compatibility tests in `integration-tests/compatibility` that ensure that older documents with edge values can be read by the new parser correctly.

File: sql/types/json.go (hunk 1 above)
Comment (nicktobey) [reply sql/types/json.go]: I don't think so?

This is only valid JSON if there are no more tokens to parse, so we expect to get an io.EOF, and everything else is an error. We have existing parsing tests that detect invalid JSON inputs and they pass here.

I must be missing something though, because when I asked Claude as an experiment, and Claude *also* insisted on adding a check that the token was nil.

File: enginetest/queries/queries.go (hunk 2 above)
Comment (nicktobey) [reply enginetest/queries/queries.go:8731]: Added.
